### DISCUSSION TITLE
Struct decomposer idiom

### DISCUSSION THREAD

**Discussion by mihaigalos** [rust-unofficial/patterns#367] [Ideas]
Hi,

I've seen [this](https://github.com/myOmikron/actix-toolbox/blob/main/src/oidc/config.rs#LL69C9-L80C18) idiom before and thought maybe we could incorporate it into this repo:

```rust
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Config {
    pub finish_login_url: RedirectUrl,
    pub post_auth_url: String,
    pub provider: Provider,
    pub scopes: HashSet<Scope>,
    pub session_keys: SessionKeys,
}

impl Config {
    pub async fn discover(self) -> Result<Data<Client>, DiscoveryError<HttpClientError>> {
        let Config {
            finish_login_url,
            post_auth_url,
            provider:
                Provider {
                    client_id,
                    client_secret,
                    discover_url,
                },
            scopes,
            session_keys,
        } = self;
        let provider_metadata = CoreProviderMetadata::discover_async(discover_url, async_http_client).await?;
        let client = CoreClient::from_provider_metadata(provider_metadata, client_id, client_secret)
                         .set_redirect_uri(finish_login_url);
```
Creating an anonymous Config object decomposes `self` into the structure fields.
This essentially eliminates the need for `self.` every time we access a member variable.

**Comment by simonsan** [comment]
It's a nice feature for sure, but as it's already in the [book](https://doc.rust-lang.org/book/ch18-03-pattern-syntax.html#destructuring-structs) I'm not sure if it would make sense to write something here which wouldn't give more information about it essentially.

**Reply by mihaigalos** [reply]
I don't remember reading this a couple of years back when I read the book. Seems they're named "Destructing Structs". Ok to skip then.
